### PR TITLE
Handle F# lists that implement IReadOnlyCollection<T> (F# 4.1 and above)

### LINKED ISSUE #1180: Support upcoming F# 4.1
There is a [breaking change](https://github.com/Microsoft/visualfsharp/issues/2257) in F# 4.1 concerning Json.NET.
It causes an exception when deserializing to FSharpList type

_"Newtonsoft.Json.JsonSerializationException: Cannot create and populate list type Microsoft.FSharp.Collections.FSharpList`1[System.Int32]. Path '', line 1, position 1."_

Because in F# 4.1 FSharp list type now [implements IReadOnlyCollection](https://github.com/Microsoft/visualfsharp/commit/13b0ba050d450ca09a3e461d6355e490a273fca9)

this conditional block in [JsonArrayContract.cs](https://github.com/JamesNK/Newtonsoft.Json/blob/master/Src/Newtonsoft.Json/Serialization/JsonArrayContract.cs#L182)
```csharp
#if HAVE_READ_ONLY_COLLECTIONS
else if (ReflectionUtils.ImplementsGenericDefinition(underlyingType, typeof(IReadOnlyCollection<>), out tempCollectionType))
```
is executed instead of [the following](https://github.com/JamesNK/Newtonsoft.Json/blob/master/Src/Newtonsoft.Json/Serialization/JsonArrayContract.cs#L210)
```csharp
#if HAVE_FSHARP_TYPES
if (!HasParameterizedCreatorInternal && underlyingType.Name == FSharpUtils.FSharpListTypeName)
```

Basically as for now the logic in JsonArrayContract assumes FSharpList type doesn't implement IReadOnlyCollection. This is no longer true in the upcoming F# release.

### PR DESCRIPTION
**Pull request by fgreinacher**
Fixes #1180.

### CODE REVIEW

**Comment by fgreinacher** [comment]
@JamesNK Is there any way to get a nupkg from a PR so that @majocha can verify?

**Comment by majocha** [comment]
I've built Newtonsoft.Json.dll from your PR. Works fine!

**Comment by JamesNK** [comment]
Hrm, it is unfortunate that there will be no way to unit test that both old and new FSharp library users will both work.

File: Src/Newtonsoft.Json/Serialization/JsonArrayContract.cs [hunk 1]
```diff
@@ -191,6 +191,15 @@ public JsonArrayContract(Type underlyingType)
 
                 _genericCollectionDefinitionType = typeof(List<>).MakeGenericType(CollectionItemType);
                 _parameterizedConstructor = CollectionUtils.ResolveEnumerableCollectionConstructor(CreatedType, CollectionItemType);
+
+#if HAVE_FSHARP_TYPES
+                if (!HasParameterizedCreatorInternal && underlyingType.Name == FSharpUtils.FSharpListTypeName)
```
Comment (JamesNK) [inline Src/Newtonsoft.Json/Serialization/JsonArrayContract.cs]: Move this into a method that is called here and below on line 219.

**Comment by fgreinacher** [comment]
Updated commit, AppVeyor seems to be hung :/

**Comment by 0x53A** [comment]
> Hrm, it is unfortunate that after this there will be no way to unit test that users of both old and new FSharp library will work.

Hm I think it should be possible to check-in the binaries of both FSharp.Core.dlls, and then reference both with a different alias (https://msdn.microsoft.com/de-de/library/ms173212.aspx). You would need to rename the filename of one of them though ...

File: Src/Newtonsoft.Json/Serialization/JsonArrayContract.cs [hunk 2]
```diff
@@ -207,11 +212,7 @@ public JsonArrayContract(Type underlyingType)
                 _parameterizedConstructor = CollectionUtils.ResolveEnumerableCollectionConstructor(underlyingType, CollectionItemType);
 
 #if HAVE_FSHARP_TYPES
-                if (!HasParameterizedCreatorInternal && underlyingType.Name == FSharpUtils.FSharpListTypeName)
```
Comment (dsyme) [inline Src/Newtonsoft.Json/Serialization/JsonArrayContract.cs]: You will still need this when using FSharp.Core 4.4.1.0 and before.  FSharp.Core only impemented IEnumerable up to version 4.4.1.0 inclusive

Comment (fgreinacher) [inline Src/Newtonsoft.Json/Serialization/JsonArrayContract.cs]: Yes, I extracted the logic in a new method and call it from here and from the other branch.

**Comment by JamesNK** [comment]
LGTM

I'll probably do a new pre-release version on NuGet sometime soon so this is available for people who use the latest version of FSharp.Core

**Comment by dsyme** [comment]
@JamesNK Thank you

**Comment by blumu** [comment]
This bug also affects FSharpLu.Json (https://github.com/Microsoft/fsharplu/issues/14). Is the new Nuget available for testing now? @JamesNK

**Comment by majocha** [comment]
it's fixed in beta
[https://www.nuget.org/packages/Newtonsoft.Json/10.0.1-beta1](https://www.nuget.org/packages/Newtonsoft.Json/10.0.1-beta1)

**Pull request merged**